node/src/config_files/mod.rs

```diff
@@ -40,58 +35,62 @@ use self::{
 };
 
 /// The node configuration.
+#[must_use]
 #[derive(Serialize, Deserialize, Debug)]
 pub struct NodeConfigFile {
-    /// The path to the data directory.
-    ///
-    /// By default the config file is created inside of the data directory.
-    pub datadir: PathBuf,
-
     // Subsystems configurations.
-    pub chainstate: ChainstateLauncherConfigFile,
-    pub p2p: P2pConfigFile,
-    pub rpc: RpcConfigFile,
+    pub chainstate: Option<ChainstateLauncherConfigFile>,
+    pub p2p: Option<P2pConfigFile>,
+    pub rpc: Option<RpcConfigFile>,
 }
 
 impl NodeConfigFile {
     /// Creates a new `Config` instance with the given data directory path.
-    pub fn new(datadir: PathBuf) -> Result<Self> {
-        let chainstate = ChainstateLauncherConfigFile::new();
-        let p2p = P2pConfigFile::default();
-        let rpc = RpcConfigFile::default();
+    pub fn new() -> Result<Self> {
         Ok(Self {
-            datadir,
-            chainstate,
-            p2p,
-            rpc,
+            chainstate: None,
+            p2p: None,
+            rpc: None,
         })
     }
 
+    fn read_to_string_with_policy<P: AsRef<Path>>(config_path: P) -> Result<String> {
+        let config_as_str = if config_path.as_ref().exists() {
+            if config_path.as_ref().is_file() {
```

Comment:
Why is this check is important? If it is only needed for a better error reporting than I think that the standard errors are pretty descriptive (something like `kind: IsADirectory, message: "Is a directory"`).

Reply:
I think there's a merit to mentioning the error is related to the config file. Maybe it is possible to attach extra info to the default error message.

Reply:
Correct. There's merit in that, it's just difficult to do both (write OS message + detect) and it'll be considered code bloat in a way. This error will almost never happen, so we catch it just in case and move on.

Reply:
I still don't get it. 🙃  I think that the message produced by the following code is enough to understand the root of the issue:
```rust
    fn read_to_string_with_policy<P: AsRef<Path>>(config_path: P) -> Result<String> {
        let config_as_str = if config_path.as_ref().exists() {
            fs::read_to_string(config_path.as_ref()).context(format!(
                "Unable to read config file in {}",
                config_path.as_ref().display()
            ))?
        } else {
            "".into()
        };
        Ok(config_as_str)
    }
```

By the way, the `is_file` can return false because of a permission error and in that case the current message will be misleading.

Reply:
Alright. Went with that.